SimPEG/Mesh/CylMesh.py

```diff
@@ -322,14 +321,113 @@ def aveF2CCV(self):
         if getattr(self, '_aveF2CCV', None) is None:
             n = self.vnC
             if self.isSymmetric:
-                avR = av(n[0])[:,1:]
+                avR = Utils.av(n[0])[:,1:]
                 avR[0,0] = 1.
-                self._aveF2CCV = sp.block_diag((sp.kron(speye(n[2]), avR),
-                                                sp.kron(av(n[2]), speye(n[0]))), format="csr")
+                self._aveF2CCV = sp.block_diag((sp.kron(Utils.speye(n[2]), avR),
+                                                sp.kron(Utils.av(n[2]), Utils.speye(n[0]))), format="csr")
             else:
                 raise NotImplementedError('wrapping in the averaging is not yet implemented')
         return self._aveF2CCV
 
+
+    def _fastInnerProduct(self, projType, prop=None, invProp=False, invMat=False):
+        """
+            Fast version of getFaceInnerProduct.
+            This does not handle the case of a full tensor prop.
+
+            :param numpy.array prop: material property (tensor properties are possible) at each cell center (nC, (1, 3, or 6))
+            :param str projType: 'E' or 'F'
+            :param bool returnP: returns the projection matrices
+            :param bool invProp: inverts the material property
+            :param bool invMat: inverts the matrix
+            :rtype: scipy.sparse.csr_matrix
+            :return: M, the inner product matrix (nF, nF)
+        """
+        assert projType in ['F', 'E'], "projType must be 'F' for faces or 'E' for edges"
+
+        if prop is None:
+            prop = np.ones(self.nC)
+
+        if invProp:
+            prop = 1./prop
+
+        if Utils.isScalar(prop):
+            prop = prop*np.ones(self.nC)
+
+        if prop.size == self.nC:
+            Av = getattr(self, 'ave'+projType+'2CC')
+            Vprop = self.vol * Utils.mkvc(prop)
+            M = self.dim * Utils.sdiag(Av.T * Vprop)
+        elif prop.size == self.nC*self.dim:
+            Av = getattr(self, 'ave'+projType+'2CCV')
+            V = sp.kron(sp.identity(self.dim), Utils.sdiag(self.vol))
+            M = Utils.sdiag(Av.T * V * Utils.mkvc(prop))
+        else:
+            return None
+
+        if projType == 'E':
+                M = 0.5*M
```

Comment:
for the edge inner products to function properly (by this I mean, have the EM analytics pass), we need this factor of 0.5 due to the bug fix in the averaging. It is not obvious to me why we need this - @rowanc1 : can you shed some light here?